### PR TITLE
Optimising and swipe delete

### PR DESCRIPTION
**Pull request by inetamagone**
Optimising and swipe delete

### CODE REVIEW

File: To-do-list/AppDelegate.swift [hunk 1]
```diff
@@ -11,40 +11,48 @@ import UIKit
...
+        let viewModel = ViewModel()
+        viewModel.onOpenTaskViewController = { [ weak self ] in
+            guard let self = self else { return }
+            let taskController = self.createTaskViewController(controller: controller)
+            self.navigationController?.pushViewController(taskController, animated: true)
         }
-        controller.configure(mainViewModel: mainViewModel)
-        
+        viewModel.onDeleteAll = { [ weak self ] in
+            controller.reloadTableView()
```
Comment (eugenemokhan) [inline To-do-list/AppDelegate.swift]: should remove this logic, can override some closures at viewController

File: To-do-list/Controller/ViewController.swift [hunk 2]
```diff
@@ -23,43 +23,90 @@ class ViewController: UIViewController {
...
     @objc func addTask() {
-        mainViewModel?.shouldOpenTaskViewController()
+        viewModel?.shouldOpenTaskViewController()
     }
     
     @objc func deleteTasks() {
-        mainViewModel?.deleteAllTasks()
-        mainViewModel?.shouldDeleteTaskManagerTasks()
-        reloadTableView()
+        deleteAlertMessage()
```
Comment (eugenemokhan) [inline To-do-list/Controller/ViewController.swift]: you call a function from function, think you can migrate the logic from deleteAlertMessage()

File: To-do-list/Controller/ViewController.swift [hunk 3]
```diff
@@ -23,43 +23,90 @@ class ViewController: UIViewController {
...
         let trashItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.trash, target: self, action: #selector(deleteTasks))
-        trashItem.tintColor = .black
+        trashItem.tintColor = .systemIndigo
         self.navigationItem.leftBarButtonItem = trashItem
     }
+    
+    func deleteAlertMessage() {
+        if viewModel?.taskCount() != 0 {
+            let alertController = UIAlertController(title: "Delete All tasks?", message: "Do you want to delete them all?", preferredStyle: .actionSheet)
+            viewModel?.setupAlertMessage(alertController: alertController)
```
Comment (eugenemokhan) [inline To-do-list/Controller/ViewController.swift]: configure your alert at viewContorller, not at viewModel

File: To-do-list/Controller/ViewController.swift [hunk 4]
```diff
@@ -23,43 +23,90 @@ class ViewController: UIViewController {
...
         
         return cell
     }
+    
+    // Message appears when there are no tasks
+    func setEmptyView() {
+        let emptyView = UIView(frame: CGRect(x: tableView?.center.x ?? 0, y: tableView?.center.y ?? 0, width: tableView?.bounds.size.width ?? 0, height: tableView?.bounds.size.height ?? 0))
+        let titleLabel = UILabel()
+        let messageLabel = UILabel()
+        viewModel?.setEmptyViewAppearence(emptyView: emptyView, titleLabel: titleLabel, messageLabel: messageLabel)
```
Comment (eugenemokhan) [inline To-do-list/Controller/ViewController.swift]: configure your emptyView, titleLabel and messageLabel at viewController, not at viewModel

File: To-do-list/Controller/ViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        if taskManager.tasks[index].completed == false {
+            taskManager.tasks[index].completed = true
+        } else {
+            taskManager.tasks[index].completed = false
+        }
+        return taskManager.tasks[index].completed
+    }
+    
+    func setEmptyViewAppearence(emptyView: UIView, titleLabel: UILabel, messageLabel: UILabel) {
+        titleLabel.translatesAutoresizingMaskIntoConstraints = false
```
Comment (eugenemokhan) [inline To-do-list/Controller/ViewModel.swift]: remove all ui logic from viewModel

File: To-do-list/Controller/ViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        titleLabel.text = "You don't have any tasks"
+        messageLabel.textColor = UIColor.gray
+        messageLabel.font = UIFont(name: "Helvetica Neue", size: 17)
+        messageLabel.text = "You can add tasks by pressing on the +  button"
+        messageLabel.numberOfLines = 0
+        messageLabel.textAlignment = .center
+    }
+    
+    func restoreView(tableView: UITableView) {
+        tableView.backgroundView = nil
```
Comment (eugenemokhan) [inline To-do-list/Controller/ViewModel.swift]: remove all ui logic from viewModel

File: To-do-list/Controller/ViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    
+    func setupAlertMessage(alertController: UIAlertController) {
+        let addActionButton = UIAlertAction(title: "Delete", style: .destructive) { action in
```
Comment (eugenemokhan) [inline To-do-list/Controller/ViewModel.swift]: remove all ui logic from viewModel

**Review by eugenemokhan — CHANGES_REQUESTED** [review]

File: To-do-list/AppDelegate.swift (hunk 1 above)
Comment (inetamagone) [reply To-do-list/AppDelegate.swift]: Ok

File: To-do-list/Controller/ViewController.swift (hunk 2 above)
Comment (inetamagone) [reply To-do-list/Controller/ViewController.swift]: Function mainViewModel.deleteAllTasks() should still stay in alert message? It is alert message Delete button action. I can migrate only if..else condition

File: To-do-list/Controller/ViewController.swift (hunk 3 above)
Comment (inetamagone) [reply To-do-list/Controller/ViewController.swift]: Ok

File: To-do-list/Controller/ViewController.swift (hunk 4 above)
Comment (inetamagone) [reply To-do-list/Controller/ViewController.swift]: Ok

File: To-do-list/Controller/ViewModel.swift (hunk 5 above)
Comment (inetamagone) [reply To-do-list/Controller/ViewModel.swift]: Ok

File: To-do-list/Controller/ViewModel.swift (hunk 6 above)
Comment (inetamagone) [reply To-do-list/Controller/ViewModel.swift]: Ok

File: To-do-list/Controller/ViewController.swift [hunk 8]
```diff
@@ -32,7 +32,11 @@ class ViewController: UIViewController {
     }
     
     @objc func deleteTasks() {
-        deleteAlertMessage()
+        if viewModel?.taskCount() != 0 {
+            deleteAlertMessage()
+        } else {
+            reloadTableView()
+        }
     }
```
Comment (inetamagone) [inline To-do-list/Controller/ViewController.swift:40]: Is it what you meant?

Comment (eugenemokhan) [reply To-do-list/Controller/ViewController.swift:40]: yep

File: To-do-list/AppDelegate.swift [hunk 9]
```diff
@@ -11,40 +11,57 @@ import UIKit
...
+        viewModel.onOpenTaskViewController = { [ weak self ] in
+            guard let self = self else { return }
+            let taskController = self.createTaskViewController(controller: controller)
+            self.navigationController?.pushViewController(taskController, animated: true)
         }
-        controller.configure(mainViewModel: mainViewModel)
-        
+        viewModel.onOpenEditViewController = { [ weak self ] in
+            guard let self = self else { return }
+            let taskModel = TaskViewModel()
```
Comment (eugenemokhan) [inline To-do-list/AppDelegate.swift]: where did you set taskModel to your taskController?
try to create function for this logic

**Review by eugenemokhan — CHANGES_REQUESTED** [review]

File: To-do-list/AppDelegate.swift (hunk 9 above)
Comment (inetamagone) [reply To-do-list/AppDelegate.swift]: I set a function configureTaskModel()

**Review by eugenemokhan — APPROVED** [review]

**Pull request merged**